README
Fryktelig fin faktura
=====================

To force the program to start in english, run "LC_ALL=en_US ./faktura".
Some hard coded strings might still be in norwegian, as this program is made exlusively for use in Norway.


Dette er et enkelt program for å lage fakturaer. Fakturaene lages som pdf_ og kan skrives ut på skjema f60 eller sendes som epost. Programmet lager en database over varer, kunder, og utsendte fakturaer.

Programmet startes ved å kjøre «faktura». For tekstmodus, bruk opsjonen -i. Flere opsjoner_ er også mulig.
Miljøvariabelen_ FAKTURADIR kan brukes til å styre hvor databasen skal lagres. Dette kan brukes til å lage fakturaer for ulike firmaer.
Eksempel:

$ FAKTURADIR=/var/fakturadb ./faktura

Du finner mer informasjon om programmet, og informasjon om feilsøking på prosjektsidene_.

.. _pdf: http://en.wikipedia.org/wiki/Portable_Document_Format
.. _opsjoner og miljøvariabler: https://sourceforge.net/p/finfaktura/wiki/Starte%20programmet/
.. _prosjektsidene: https://sourceforge.net/projects/finfaktura/

Endre fakturanummer
===================

Det er nå mulig å endre første fakturanummer, men det virker kun dersom du ikke allerede har laget fakturaer.
Dette kan kun gjøres én gang. Du kan ikke endre dette tallet senere.
I Oppsett-menyen har du et valg som heter «Første fakturanummer». Klikk på den og følg instruksjonene.

Diskusjonsgruppe
================

Lurer du på noe? Er det noe som mangler?
Ta en titt på diskusjonsgruppa, hvor nye versjoner og funksjoner blir diskutert, hvor problemer blir løst, og hvor solen aldri går ned:
https://lists.sourceforge.net/lists/listinfo/finfaktura-support

Vi har også en kanal på Libera: #finfaktura

Systemkrav
==========

Programmet er laget for å kjøre på GNU/Linux, Windows og Apple osx, men utviklerteamet bruker og tester kun i GNU/Linux. Vi kan derfor ikke garantere at det fungerer i andre operativsystemer.
Fryktelig fin faktura utvikles og testes i Mageia med python 2.7.
Programmet finnes foreløpig ikke i python3-utgave, da ingen av utviklerne egentlig kan python-programmering. :-)

Følgende moduler er påkrevet:

- python_ 3, versjon 3.8.10 eller nyere.
- python-modulen PyQt5_, versjon 5.14.1 eller nyere.
- python-modulen Reportlab_, versjon 3.5.34 eller nyere.

Disse oppfylles automatisk når du installerer fra deb/rpm-pakke.

.. _python: http://python.org/
.. _PyQt5: http://www.riverbankcomputing.com/software/pyqt/intro
.. _reportlab: http://www.reportlab.org/rl_toolkit.html


Windows
-------

Det finnes ingen oppdatert Windows-installasjonsfil, men sjekk ut
versjon 6a8428c9158106936366699fc303e510fb9437a4 og les denne filen
for hvordan du får det til å kjøre med en eldre utgave.


Opphavsrett
===========

Idé og koding: Håvard Gulldahl <havard@gulldahl.no>
Dette er Fri programvare, du kan fritt gjenbruke og videredistribuere under GPL2-lisensen.
Ny prosjektledelse: Johnny A. Solbu <johnny@solbu.net> fra 2015

Se LICENSE eller http://www.gnu.org/licenses/old-licenses/gpl-2.0.html#TOC1


Kjente feil og planlagte egenskaper
===================================

Se https://sourceforge.net/p/finfaktura/tickets/


debian/changelog
finfaktura (3.0.0-1) stable; urgency=low

  * Ny versjon
  * LOCALE-fikser
  * Dokumenter installasjonsprosedyre på Windows
  * Python2 -> python3
  * PyQT4 -> PyQT5

 -- Johnny A. Solbu <johnny@solbu.net>  Fri, 25 Jun 2021 15:48:04 +0200

finfaktura (2.2.1-1) stable; urgency=low

  * Ny versjon
  * Fiks PIL-importering (Bugrapport via IRC)

 -- Johnny A. Solbu <johnny@solbu.net>  Mon, 18 Jun 2018 21:52:24 +0100

finfaktura (2.2.0-1) stable; urgency=low

  * Ny versjon
  * Legg til Gui-logo

 -- Johnny A. Solbu <johnny@solbu.net>  Wed, 25 Feb 2018 00:18:20 +0100

finfaktura (2.1.4-1) stable; urgency=medium

  * Ny versjon
  * Få tilbake «KVITTERING» på kvitteringsutskrifter

 -- Johnny A. Solbu <johnny@solbu.net>  Wed, 07 Jun 2017 16:05:22 +0200

finfaktura (2.1.3-1) stable; urgency=low

  * Ny versjon

 -- Johnny A. Solbu <johnny@solbu.net>  Thu, 17 Nov 2016 00:09:18 +0100

finfaktura (2.1.2-1) stable; urgency=medium

  * Ny versjon
  * Støtte for oversettelser

 -- Johnny A. Solbu <johnny@solbu.net>  Fri, 06 Nov 2015 01:51:01 +0100

finfaktura (2.1.1-1) stable; urgency=medium

  * Ny versjon

 -- Johnny A. Solbu <johnny@solbu.net>  Fri, 30 Oct 2015 02:25:26 +0100

finfaktura (2.1.0-1) stable; urgency=low

  * Ny versjon
  * Støtte for oversettelser

 -- Johnny A. Solbu <johnny@solbu.net>  Sat, 24 Oct 2015 02:03:19 +0200

finfaktura (2.0.9-1) stable; urgency=low

  * Ny versjon
  * Ny maintainer

 -- Johnny A. Solbu <johnny@solbu.net>  Mon, 17 Aug 2015 17:27:45 +0200

finfaktura (2.0.6b) karmic; urgency=low

  * Fikser installasjon på systemer som har både python2.5 og 2.6 installert (bl.a. Ubuntu 9.04)

 -- Håvard Gulldahl <havard@gulldahl.no>  Mon, 18 May 2009 12:17:22 +0200

finfaktura (2.0.6) jaunty; urgency=low

  * f60.py: fikser bug#54, hvor linjeskift ble til sort firkant -- takk,
    Harald!
  * f60.py: fikser bug#56, beløp skrives med $-tegn dersom locale ikke er
    riktig -- takk igjen, Harald!
  * f60.py: versjon 0.16
  * fikser bug#55, hvor ikke-ascii tegn i status-felt på kunde eller vare
    førte til problemer -- takk, Johan!
  * småendringer i gui for å gjøre den enklere å bruke på små skjermer
  * bedre installasjonsinstillinger for debian-derivater og
    netinstaller for windows

 -- Håvard Gulldahl <havard@gulldahl.no>  Mon, 13 Apr 2009 22:44:48 +0200

finfaktura (2.0.5) jaunty; urgency=low

  * f60.py: massive forbedringer av blanketten (stor takk til cbrattli),
     se issue#38
  * f60.py: skriver beløp i kroner og med kommaskilletegn
  * f60.py: Fiks i mod10-algoritme, og inn med mod11 (KID) - igjen, takk til cbrattli. 
  * f60.py: Spesifisert momsgrunnlag
  * fiks utskrift av kontonummer som starter på 0 (issue#47 igjen)
  * Laster logoer på en annen måte (forhåpentlig løser dette issue#48)

 -- Håvard Gulldahl <havard@gulldahl.no>  Thu, 12 Mar 2009 23:15:01 +0100

finfaktura (2.0) intrepid; urgency=low

  * Fikser rapport.py: utskrift av pdf og rapport fra firma med '&' i
    firmanavnet (issue#43) -- takk, Eivind!
  * Fikser scrollelinjer dersom skjermen/vinduet er mindre enn informasjonen
    som skal vises -- takk, Johan!
  * Bedre installasjon på Debian/Ubuntu
  * Fikser vising av PDF på windows

 -- Håvard Gulldahl <havard@gulldahl.no>  Wed, 03 Dec 2008 01:33:38 +0100

finfaktura (1.9.9b3-3) intrepid; urgency=low

  * Sørger for at scripts/forbered_ressurser.sh blir kjørt, og 
    .ui- og .qrc-filer blir med i .deb-pakken

 -- Håvard Gulldahl <havard@gulldahl.no>  Tue, 28 Oct 2008 16:15:25 +0100

finfaktura (1.9.9b3-2) intrepid; urgency=low

  * Fikser gui_sikkerhetspkopi.py (issue#36) - takk, Johan!

 -- Håvard Gulldahl <havard@gulldahl.no>  Sun, 26 Oct 2008 02:31:49 +0200

finfaktura (1.9.9b3) intrepid; urgency=low

  * Mer bugfiksing på epostsending
  * Gir navn på vinduer
  * Meny-oppføring

 -- Håvard Gulldahl <havard@gulldahl.no>  Wed, 22 Oct 2008 00:25:41 +0200

finfaktura (1.9.9b2) intrepid; urgency=low

  * Bugfiks: epostsending, firmainfo/logo, ny vare, eposttest ++
  * Eksterne programmer startes på en universell måte som funker på alle
    plattformer
  * Logoer og dokumenter installeres
  * Installeres med meny-oppføring (issue#1)

 -- Håvard Gulldahl <havard@gulldahl.no>  Mon, 13 Oct 2008 00:21:39 +0200

finfaktura (1.9.9b) intrepid; urgency=low

  * Grensesnittet er portert til PyQt4 (issue#3) og redesignet
  * Fikser justering av utskrift på skjema F60 (issue#30)
  * Det går an å føre halve antall av varer (f.eks. timer) (issue#22)
  * Kan sette vilkårlig forfallsdato på ny faktura (issue#25)
  * Fikser en rekke bugs, for eksempel issue#32
  * En rekke egne vinduer introduseres, bl.a. Om.. og Lisens.. -vinduer
  * Har tatt ulike grep for å å gjøre programvinduet så lavt som mulig (s.a.
    det skal funke på små skjermer/oppløsninger)

 -- Håvard Gulldahl <havard@gulldahl.no>  Thu, 04 Sep 2008 01:02:39 +0200

finfaktura (1.3.2) intrepid; urgency=low

  * Bug-en ved første gangs oppstart var ikke helt skvist

 -- Håvard Gulldahl <havard@gulldahl.no>  Sun, 11 May 2008 11:00:50 +0200

finfaktura (1.3.1) intrepid; urgency=low

  * Viktig bugfiks for uventet inndata

 -- Håvard Gulldahl <havard@gulldahl.no>  Fri, 09 May 2008 21:02:17 +0200

finfaktura (1.3) intrepid; urgency=low

  * Fikset layout - den er justerbar nå
  * Kan skrive ut rapporter
  * Justerbare plasseringer for PDF-programmer

 -- Håvard Gulldahl <havard@lurtgjort.no>  Fri, 09 May 2008 20:23:55 +0200

finfaktura (1.1) stable; urgency=low

  * Fikset bug hvor nye varer ikke kunne legges til på direkten
  * Ny-faktura-vindu: Varens pris og mva blir oppdatert når man blar gjennom lista
  * Restrukturert kode for bedre layout
  * Kan angre på at en faktura er betalt
  * Test epostoppsett funker igjen
  * Mer hjelpetekst

 -- Håvard Dahle <havard@lurtgjort.no>  Thu,  3 May 2007 02:01:40 +0200

finfaktura (1.0) stable; urgency=low

  * Ny og bedre dialog for å legge inn ny faktura
  * Forbedret epost-funksjonalitet: kan bruke smtp.gmail.com:587

finfaktura/rapport.py
# -*- coding: utf-8 -*-
###########################################################################
#    Copyright (C) 2005-2009 Håvard Gulldahl
#    <havard@lurtgjort.no>
#
#    Lisens: GPL2
#
# $Id: oppgradering.py 217 2007-05-02 23:25:16Z havard.dahle $
###########################################################################

import time
import os
import logging

from . import fakturafeil, fil

try:
    import reportlab
    from reportlab.lib.styles import getSampleStyleSheet
    from reportlab.lib.pagesizes import A4
    from reportlab.platypus import Paragraph, SimpleDocTemplate
    REPORTLAB=True
except ImportError:
    REPORTLAB=False
    #raise

class rapport:
    'Lager økonomisk rapport på pdf'
    oppdatert = False

    def __init__(self, filnavn=None, rapportinfo={}):
        if not REPORTLAB:
            raise fakturafeil.InstallasjonsFeil('python-reportlab er ikke installert. Kan ikke lage PDF!')
        if filnavn is None:
            import tempfile
            f, self.filnavn = tempfile.mkstemp(suffix='.pdf', prefix='rapport-')
            os.close(f) # lukk fil-deskriptoren
        else:
            self.filnavn = filnavn
        self.info = rapportinfo
        self.stiler = getSampleStyleSheet()
        self.normal = self.stiler['BodyText']
        self.liste  = self.stiler['Bullet']
        self.overskrift = self.stiler['Heading3']
        self.seksjonover = self.stiler['Heading2']
        #print dir(self.seksjonover)
        self.tittel = self.stiler['Heading1']
        self.flow = []
        self.flow.append(Paragraph('Fakturaer hos %s' % self.tryggXml(self.info['firma'].firmanavn), self.tittel))
        self.flow.append(Paragraph('Generert av <i>Fryktelig Fin Faktura</i> den %s' % time.strftime("%Y-%m-%d", time.localtime()), self.normal))
        det = 'Viser fakturaer '
        if self.info['visubetalte']: det += '(også ubetalte) '
        else: det += '(ikke ubetalte) '
        if self.info['dato'] != (None, None):
            _fra, _til = self.info['dato']
            if _fra is not None: det += "fra %s" % time.strftime("%Y-%m", time.localtime(_fra))
            if _til is not None: det += "til %s" % time.strftime("%Y-%m", time.localtime(_til))
        if self.info['kunde'] is not None:
            det += 'sendt til %s' % self.info['kunde'].navn
        self.flow.append(Paragraph(det, self.normal))
        self.okonomi = {'inn':0.0, 'mva':0.0, 'b':0, 'u':0}
        self.seksjon = ''
        logging.debug(self.info)

    def lag(self):
        dok = SimpleDocTemplate(self.filnavn,pagesize = A4)
        #dok.setAuthor("%s (Fryktelig Fin Faktura)" % self.info.firma.kontaktperson)
        #dok.setTitle("Fakturaer hos %s" % self.info.firma.firmanavn)
        #canvas.setSubject("How to Generate PDF files using the ReportLab modules")
        dok.build(self.flow)
        self.oppdatert = True

    def vis(self):
        logging.debug("viser rapport '%s'", self.filnavn)
        if not self.oppdatert:
            self.lag()
        return fil.vis(self.filnavn)

    def lastOrdreliste(self, ordreliste):
        for o in ordreliste: self.leggTilOrdre(o)

    def leggTilOrdre(self, ordre):
        self.oppdatert = False
        status = ""
        if self.info['sortering'] is not None:
            if self.info['sortering'] == 'dato':
                _seksjon = time.strftime("%Y-%m", time.localtime(ordre.ordredato))
            elif self.info['sortering'] == 'kunde':
                _seksjon = ordre.kunde.navn
            if self.seksjon != _seksjon:
                self.flow.append(Paragraph(_seksjon, self.seksjonover))
                self.seksjon = _seksjon
        if ordre.kansellert:
            status = "<font color=red><b>kansellert</b></font>"
        elif ordre.betalt:
            status = "<font color=green>betalt</font>"
            self.okonomi['inn'] += ordre.finnPris()
            self.okonomi['mva'] += ordre.finnMva()
            self.okonomi['b'] += 1
        else:
            status = "<font color=red>ubetalt</font>"
            self.okonomi['u'] += 1
        self.flow.append(Paragraph("ordre <i># %04i</i> (%s), utformet til %s den %s\n" % (ordre._id, status, ordre.kunde.navn, time.strftime("%Y-%m-%d", time.localtime(ordre.ordredato))), self.overskrift))
        if ordre.linje:
            for vare in ordre.linje:
                self.flow.append(Paragraph("<li> #%i: %s </li>" % (vare._id, str(vare)), self.liste))

    def tryggXml(self, s):
      return s.replace('&', '&amp;').replace('<', '&lt;')
